### PR TITLE
verify cached data

### PR DESCRIPTION
**Pull request by schrodit**
**What this PR does / why we need it**:

Locally cached blobs are now validated with their digest.
Tampered blob are deleted and a missed cache reported.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature user
Locally cached blobs are now verified to detect tampered or invalid data.
```

### CODE REVIEW

File: ociclient/cache/cache.go [hunk 1]
```diff
@@ -176,9 +176,11 @@ func (lc *layeredCache) get(dgst string, desc ocispecv1.Descriptor) (os.FileInfo
 		}
 		return nil, nil, err
 	}
-	// do simple blob verification
-	// todo: use digest to check if blob is valid
-	if info.Size() != desc.Size {
+	verified, err := verifyBlob(lc.baseFs.FileSystem, info, dgst, desc)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to verify blob: %w", err)
```
Comment (achimweigel) [inline ociclient/cache/cache.go:181]: Could there be an endless loop if blob in cache is damaged?

File: ociclient/cache/cache.go [hunk 2]
```diff
@@ -230,9 +232,11 @@ func (lc *layeredCache) getFromOverlay(dgst string, desc ocispecv1.Descriptor) (
 		return nil, nil, ErrNotFound
 	}
 
-	// do simple blob verification
-	// todo: use digest to check if blob is valid
-	if info.Size() != desc.Size {
+	verified, err := verifyBlob(lc.overlayFs.FileSystem, info, dgst, desc)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to verify blob: %w", err)
```
Comment (achimweigel) [inline ociclient/cache/cache.go:237]: Could there be an endless loop if blob in cache is damaged?

File: ociclient/cache/cache.go [hunk 3]
```diff
@@ -246,6 +250,26 @@ func (lc *layeredCache) getFromOverlay(dgst string, desc ocispecv1.Descriptor) (
 	return info, file, err
 }
 
-func path(desc ocispecv1.Descriptor) string {
+// verifyBlob validates that the digest of a blob
```
Comment (achimweigel) [inline ociclient/cache/cache.go:253]:
```suggestion
// verifyBlob validates the digest of a blob
```

**Review by achimweigel — CHANGES_REQUESTED** [review]

File: ociclient/cache/cache.go (hunk 1 above)
Comment (schrodit) [reply ociclient/cache/cache.go:181]: depends on the user of the cache.
If the user ignores the error and defaults to fetch from the remote repo then we will not have a endlessloop.

In our implementation I think we have that behavior that we log the error but always default to the remote repository

File: ociclient/cache/cache.go (hunk 2 above)
Comment (schrodit) [reply ociclient/cache/cache.go:237]: same as above

**Review by achimweigel — APPROVED** [review]
/lgtm

**Pull request merged**